### PR TITLE
fix(designer-ui): Prevent expressions using `&...;` syntax disappearing from HTML editor

### LINKED ISSUE #3741: Expressions containing HTML '&...;' components disappear from HTML editor
### Describe the Bug with repro steps

1. Open a workflow.
2. Add an action that uses HTML editor, such as sending an email.
3. Using the token picker, add an expression to the HTML editor that contains the text: `concat('&lt;', 'a')`.
4. Select a different node.
5. Select the edited node again.

Expected: Nothing should change between steps 3 and 5.
Actual: The expression token disappears.

### What type of Logic App Is this happening in?

Power Automate

### Are you using new designer or old designer

New Designer

### Did you refer to the TSG before filing this issue? https://aka.ms/lauxtsg

Yes

### Workflow JSON

_No response_

### Screenshots or Videos

disappearing-token

### Browser

Windows 11 / Chrome 119

### Additional context

Looks like it just completely drops the token:

image

This is likely because `nodeMap` keys are not encoded, while the HTML itself is, so the lookup doesn't work:

image

AB#25936574

### PR DESCRIPTION
**Pull request by ek68794998**
- **Please check if the PR fulfills these requirements**

* [x] The commit message follows our guidelines
* [x] Tests for the changes have been added (for bug fixes/features)
* [ ] Docs have been added / updated (for bug fixes / features)

- **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

Bug fix

- **What is the current behavior?** (You can also link to an open issue here)

When entering into the HTML editor an expression that contains an HTML entity such as `&lt;`, it will disappear from the editor upon save or after switching nodes. Fixes #3741.

- **What is the new behavior (if this is a feature change)?**

Expression no longer disappears.

- **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

No

- **Please Include Screenshots or Videos of the intended change**:

Showing token not disappearing:

disappearing-expr-fix

When the above example is saved, it properly serializes when being sent to the server:

image

AB#25970447

### CODE REVIEW

File: libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts [hunk 1]
```diff
@@ -33,7 +34,11 @@
   const nodeMap = new Map<string, ValueSegment>();
 
   const stringValue = convertSegmentsToString(value, nodeMap);
-  const dom = parser.parseFromString(stringValue, 'text/html');
+  const encodedStringValue = tokensEnabled ? stringValue.replace(/\$\[(.*?)\]\$/g, (_match, content) => {
+    return `$[${encodeSegmentValue(content)}]$`;
+  }) : stringValue;
```
[the remark this answers is not in the document]
Comment (ek68794998) [reply libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts:39]: I can convert this from a regex to a helper method if this vulnerability is a concern (which I believe it probably is).

File: libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts [hunk 2]
```diff
@@ -99,12 +104,20 @@ const appendChildrenNode = (
   // if is a text node, parse for tokens
   if ($isTextNode(childNode)) {
     const textContent = childNode.getTextContent();
+    const decodedTextContent = tokensEnabled ? decodeSegmentValue(textContent) : textContent;
```
Comment (ek68794998) [inline libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts:107]: Appears to be working for some test cases, but this one is still not working:

```
replace(replace(replace('abc','&lt;','<'),'&gt;','>'),'&quot;','"') 
```

Looking into this.

**Comment by ek68794998** [comment]
Dropping back to a draft while I work on fixing test case (see comments).

File: libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts (hunk 2 above)
Comment (ek68794998) [reply libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts:105]: Fixed.

disappearing-quot-fix

File: libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ek68794998) [reply libs/designer-ui/src/lib/editor/base/utils/parsesegments.ts]: Converted to helper method.

**Review by Eric-B-Wu — APPROVED** [review]

**Pull request merged**